File: ruby/test/chance_test.rb
# Copyright (c) 2019 by Fred George
# May be used freely except for training; license required for training.
# @author Fred George  fredgeorge@acm.org

require 'minitest/autorun'
require 'set'
require_relative '../lib/chance'

# Confirms behavior of Chance
class ChanceTest < Minitest::Test

  CERTAIN = 1.0.chance
  LIKELY = 0.75.chance
  EQUALLY_LIKELY = 0.5.chance
  UNLIKELY = 0.25.chance
  IMPOSSIBLE = 0.0.chance

  def test_equality
    assert_equal(LIKELY, 0.75.chance)
    refute_equal(LIKELY, UNLIKELY)
    refute_equal(LIKELY, Object.new)
    refute_equal(LIKELY, nil)
  end

  def test_hash_set
    assert(Set.new([LIKELY]).include?(0.75.chance))
    assert_equal(1, Set.new([LIKELY, 0.75.chance]).count)
  end

  def test_hash
    assert_equal(LIKELY.hash, 0.75.chance.hash)
  end

  def test_not
    assert_equal(LIKELY, !UNLIKELY)
    assert_equal(LIKELY, !!LIKELY)
    assert_equal(LIKELY, UNLIKELY.not)
    assert_equal(CERTAIN, !IMPOSSIBLE)
    assert_equal(0.3.chance, !!0.3.chance)
  end

  def test_and
    assert_equal(UNLIKELY, EQUALLY_LIKELY & EQUALLY_LIKELY)
    assert_equal(0.1875.chance, LIKELY & UNLIKELY)
    assert_equal(UNLIKELY & LIKELY, LIKELY & UNLIKELY)
    assert_equal(0.09.chance, 0.9.chance & 0.1.chance)
    assert_equal(LIKELY, LIKELY & CERTAIN)
    assert_equal(IMPOSSIBLE, LIKELY & IMPOSSIBLE)
  end

end

File: csharp/OoBootCamp.Tests/ChanceTest.cs
/* 
 * Copyright (c) 2019 by Fred George
 * May be used freely except for training; license required for training.
 */

using System.Collections.Generic;
using ExtensionMethods;
using NUnit.Framework;

namespace OoBootCamp.Tests
{
    // Ensures Chance operates correctly
    [TestFixture]
    public class ChanceTest
    {
        private static readonly Chance Impossible = 0.Chance();
        private static readonly Chance Unlikely = 0.25.Chance();
        private static readonly Chance EquallyLikely = 0.5.Chance();
        private static readonly Chance Likely = 0.75.Chance();
        private static readonly Chance Certain = 1.Chance();

        [Test]
        public void Equality()
        {
            Assert.AreEqual(Likely, 0.75.Chance());
            Assert.AreEqual(Likely, 0.75.Chance());
            Assert.AreNotEqual(Likely, new object());
            Assert.AreNotEqual(Likely, null);
        }

        [Test]
        public void HashSet()
        {
            Assert.True(new HashSet<Chance>() { Likely }.Contains(0.75.Chance()));
            Assert.AreEqual(1, new HashSet<Chance>() { Likely, 0.75.Chance() }.Count);
        }

        [Test]
        public void Hash()
        {
            Assert.AreEqual(Likely.GetHashCode(), 0.75.Chance().GetHashCode());
        }

        [Test]
        public void Not()
        {
            Assert.AreEqual(Unlikely, !Likely);
            Assert.AreEqual(Likely, !!Likely);
            Assert.AreEqual(Likely, Likely.Not().Not());
            Assert.AreEqual(Impossible, !Certain);
            Assert.AreEqual(Certain, !Impossible);
            Assert.AreEqual(0.3.Chance(), !!0.3.Chance());
        }

        [Test]
        public void And()
        {
            Assert.AreEqual(Unlikely, EquallyLikely & EquallyLikely);
            Assert.AreEqual(0.1875.Chance(), Likely & Unlikely);
            Assert.AreEqual(Unlikely.And(Likely), Likely & Unlikely);
            Assert.AreEqual(Likely, Likely & Certain);
            Assert.AreEqual(Impossible, Impossible & Likely);
        }

        [Test]
        public void Or()
        {
            Assert.AreEqual(Likely, EquallyLikely | EquallyLikely);
            Assert.AreEqual(0.8125.Chance(), Likely | Unlikely);
            Assert.AreEqual(Unlikely.Or(Likely), Likely | Unlikely);
            Assert.AreEqual(Certain, Likely | Certain);
            Assert.AreEqual(Likely, Impossible | Likely);
        }
    }
}